File: watch/PadelScoreWatch/ScoreScreen.swift
import SwiftUI

/// Shared Watch control chrome. System `.bordered` buttons default to a capsule
/// on watchOS; custom score controls already use this 12pt continuous rect.
enum WatchTheme {
    static let buttonCornerRadius: CGFloat = 12

    static var buttonShape: RoundedRectangle {
        RoundedRectangle(cornerRadius: buttonCornerRadius, style: .continuous)
    }
}

extension View {
    func watchButtonBorderShape() -> some View {
        buttonBorderShape(.roundedRectangle(radius: WatchTheme.buttonCornerRadius))
    }
}

struct ScoreScreen: View {
    @EnvironmentObject private var service: MatchService
    @Environment(\.isLuminanceReduced) private var isLuminanceReduced
    let match: MatchState

    @State private var undoSide: Side?
    @State private var undoStartedAt: Date?
    @State private var undoClearTask: Task<Void, Never>?

    private var game: (left: String, right: String) { match.scoreScreenGameDisplay }
    private var games: (left: String, right: String) { match.scoreScreenSetDisplay }
    private var undoTimeout: TimeInterval { MatchSettings.quickUndoTimeoutSeconds }
    private var leftRole: String { match.servingRoleLabels.left }
    private var rightRole: String { match.servingRoleLabels.right }
    private var scoreSides: (left: Side, right: Side) { match.scoreScreenSides }

    var body: some View {
        TimelineView(
            .animation(
                minimumInterval: 1.0 / 30.0,
                paused: undoStartedAt == nil || isLuminanceReduced
            )
        ) { context in
            let progress = undoProgress(at: context.date)

            VStack(spacing: isLuminanceReduced ? 6 : 10) {
                if match.currentGame.isTieBreak {
                    tieBreakHeader
                } else if match.currentGame.isGoldenPointActive {
                    decidingPointHeader
                } else {
                    gamesHeader
                }

                HStack(spacing: 8) {
                    scoreButton(
                        logicalSide: scoreSides.left,
                        score: game.left,
                        role: leftRole,
                        tint: teamTint(for: scoreSides.left),
                        showsServeIndicator: match.currentServer == scoreSides.left,
                        progress: isLuminanceReduced ? 0 : (undoSide == scoreSides.left ? progress : 0)
                    )
                    scoreButton(
                        logicalSide: scoreSides.right,
                        score: game.right,
                        role: rightRole,
                        tint: teamTint(for: scoreSides.right),
                        showsServeIndicator: match.currentServer == scoreSides.right,
                        progress: isLuminanceReduced ? 0 : (undoSide == scoreSides.right ? progress : 0)
                    )
                }
            }
            .padding(.horizontal, 4)
            .onChange(of: progress) { _, newValue in
                if newValue >= 1 {
                    clearUndoWindow()
                }
            }
        }
        .sensoryFeedback(.impact(flexibility: .solid, intensity: 0.8), trigger: match.events.count)
        .sensoryFeedback(.impact(flexibility: .soft, intensity: 0.6), trigger: match.activeTieBreakNotice)
        .onDisappear { clearUndoWindow() }
        .onChange(of: isLuminanceReduced) { _, reduced in
            if reduced {
                clearUndoWindow()
            }
        }
    }

    private func teamTint(for logicalSide: Side) -> Color {
        logicalSide == .left ? .blue : .red
    }

    private var setScoreFont: Font {
        if isLuminanceReduced {
            return .title2.weight(.bold).monospacedDigit()
        }
        return .title3.weight(.semibold).monospacedDigit()
    }

    @ViewBuilder
    private var tieBreakHeader: some View {
        if isLuminanceReduced {
            VStack(spacing: 2) {
                Text("TB")
                    .font(setScoreFont)
                    .foregroundStyle(.primary)
                if let notice = match.activeTieBreakNotice {
                    Text(tieBreakNoticeAbbreviation(notice))
                        .font(.caption2.weight(.bold))
                        .foregroundStyle(.orange)
                }
            }
            .frame(maxWidth: .infinity)
            .accessibilityElement(children: .combine)
            .accessibilityLabel(tieBreakAccessibilityLabel)
        } else {
            VStack(spacing: 2) {
                Text("Tie-break")
                    .font(.caption.weight(.bold))
                    .foregroundStyle(.orange)
                Text("\(games.left) – \(games.right)")
                    .font(.caption2)
                    .foregroundStyle(.secondary)
                if let notice = match.activeTieBreakNotice {
                    Text(tieBreakNoticeLabel(notice))
                        .font(.caption2.weight(.semibold))
                        .foregroundStyle(.orange)
                } else {
                    Text("First to 7")
                        .font(.caption2)
                        .foregroundStyle(.secondary)
                }
            }
            .frame(maxWidth: .infinity)
            .accessibilityElement(children: .combine)
            .accessibilityLabel(tieBreakAccessibilityLabel)
        }
    }

    private var tieBreakAccessibilityLabel: String {
        var parts = ["Tie-break, 6 games all"]
        if let notice = match.activeTieBreakNotice {
            parts.append(tieBreakNoticeLabel(notice))
        }
        return parts.joined(separator: ", ")
    }

    private func tieBreakNoticeLabel(_ notice: TieBreakNotice) -> String {
        switch notice {
        case .changeServe: return "Change serve"
        case .changeSides: return "Change sides"
        }
    }

    private func tieBreakNoticeAbbreviation(_ notice: TieBreakNotice) -> String {
        switch notice {
        case .changeServe: return "Serve"
        case .changeSides: return "Sides"
        }
    }

    @ViewBuilder
    private var gamesHeader: some View {
        let status = match.settings.deuceFormat.numbersDeuceCycles ? match.gameStatusLine : nil
        let gamesLabel = "Games \(games.left) to \(games.right)"
        if isLuminanceReduced || status == nil {
            Text("\(games.left) – \(games.right)")
                .font(setScoreFont)
                .foregroundStyle(.primary)
                .accessibilityLabel(
                    status.map { "\(gamesLabel), \($0)" } ?? gamesLabel
                )
        } else {
            VStack(spacing: 2) {
                Text("\(games.left) – \(games.right)")
                    .font(setScoreFont)
                    .foregroundStyle(.primary)
                Text(status!)
                    .font(.caption2)
                    .foregroundStyle(.secondary)
            }
            .accessibilityElement(children: .combine)
            .accessibilityLabel("\(gamesLabel), \(status!)")
        }
    }

    @ViewBuilder
    private var decidingPointHeader: some View {
        let format = match.settings.deuceFormat
        let accessibility = "\(format.decidingPointLabel), next point wins"
        if isLuminanceReduced {
            Text(format.decidingPointShortLabel)
                .font(setScoreFont)
                .foregroundStyle(.primary)
                .accessibilityLabel(accessibility)
        } else {
            VStack(spacing: 2) {
                Text(format.decidingPointLabel)
                    .font(.caption.weight(.bold))
                    .foregroundStyle(.yellow)
                Text("Next point wins")
                    .font(.caption2)
                    .foregroundStyle(.secondary)
            }
            .frame(maxWidth: .infinity)
            .accessibilityElement(children: .combine)
            .accessibilityLabel(accessibility)
        }
    }

    private func scoreButton(
        logicalSide: Side,
        score: String,
        role: String,
        tint: Color,
        showsServeIndicator: Bool,
        progress: Double
    ) -> some View {
        Button {
            handleTap(logicalSide)
        } label: {
            ZStack {
                WatchTheme.buttonShape
                    .fill(tint.opacity(isLuminanceReduced ? 0.12 : 0.22))

                WatchTheme.buttonShape
                    .strokeBorder(tint.opacity(isLuminanceReduced ? 0.55 : 0.35), lineWidth: isLuminanceReduced ? 2.5 : 2)

                if progress > 0 {
                    ClockwiseRoundedRectOutline(progress: progress)
                        .stroke(
                            tint,
                            style: StrokeStyle(lineWidth: 3, lineCap: .round, lineJoin: .round)
                        )
                }

                VStack(spacing: 2) {
                    Text(score)
                        .font(isLuminanceReduced ? .title.weight(.bold).monospacedDigit() : .title2.weight(.bold).monospacedDigit())
                        .foregroundStyle(.primary)
                    if !isLuminanceReduced {
                        Text(role)
                            .font(.caption2)
                            .foregroundStyle(.secondary)
                    }
                }
                .padding(.vertical, isLuminanceReduced ? 6 : 8)

                if showsServeIndicator {
                    Image(systemName: "tennisball.fill")
                        .font(.caption2)
                        .foregroundStyle(.yellow)
                        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
                        .padding(.top, 6)
                }
            }
            .frame(maxWidth: .infinity, minHeight: isLuminanceReduced ? 64 : 72)
        }
        .buttonStyle(.plain)
        .accessibilityLabel(serveAccessibilityLabel(role: role, score: score, isServing: showsServeIndicator))
        .accessibilityHint(
            undoSide == logicalSide
                ? "Double tap again to cancel the last point"
                : "Awards a point"
        )
    }

    private func serveAccessibilityLabel(role: String, score: String, isServing: Bool) -> String {
        isServing ? "\(role) \(score), serving" : "\(role) \(score)"
    }

    private func handleTap(_ logicalSide: Side) {
        if undoSide == logicalSide {
            service.undoLastPoint()
            clearUndoWindow()
            return
        }

        let previousMatch = service.activeMatch
        service.awardPoint(to: logicalSide)
        let updatedMatch = service.activeMatch

        // For points that end a game, we show a dedicated Game interstitial with undo
        // instead of keeping the in-button quick undo active.
        if !didPointEndGame(previous: previousMatch, updated: updatedMatch) {
            beginUndoWindow(for: logicalSide)
        } else {
            clearUndoWindow()
        }
    }

    private func beginUndoWindow(for side: Side) {
        undoClearTask?.cancel()
        undoSide = side
        undoStartedAt = Date()

        let timeout = undoTimeout
        undoClearTask = Task { @MainActor in
            try? await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
            if Task.isCancelled { return }
            clearUndoWindow()
        }
    }

    private func clearUndoWindow() {
        undoClearTask?.cancel()
        undoClearTask = nil
        undoSide = nil
        undoStartedAt = nil
    }

    private func undoProgress(at date: Date) -> Double {
        guard let started = undoStartedAt, undoTimeout > 0 else { return 0 }
        return min(1, max(0, date.timeIntervalSince(started) / undoTimeout))
    }

    private func didPointEndGame(previous: MatchState?, updated: MatchState?) -> Bool {
        guard
            let previous,
            let updated,
            previous.status == .inProgress,
            updated.status == .inProgress
        else {
            return false
        }

        // A game boundary is visible as either:
        // - current set games increasing (regular game win), or
        // - completed sets increasing (set-clinching game).
        let previousGamesTotal = previous.currentSet.leftGames + previous.currentSet.rightGames
        let updatedGamesTotal = updated.currentSet.leftGames + updated.currentSet.rightGames
        if updatedGamesTotal > previousGamesTotal {
            return true
        }
        return updated.completedSets.count > previous.completedSets.count
    }
}

/// Draws a continuous rounded-rect stroke that grows clockwise from the top-center.
struct ClockwiseRoundedRectOutline: Shape {
    var progress: Double
    var cornerRadius: CGFloat = WatchTheme.buttonCornerRadius

    var animatableData: Double {
        get { progress }
        set { progress = newValue }
    }

    func path(in rect: CGRect) -> Path {
        let radius = min(cornerRadius, min(rect.width, rect.height) / 2)
        let full = Path(roundedRect: rect, cornerRadius: radius, style: .continuous)
        let fraction = max(0, min(1, progress))
        guard fraction > 0 else { return Path() }
        if fraction >= 1 { return full }

        let start = Self.topCenterStartFraction(rect: rect, radius: radius)
        return Self.trimmedWrapping(full, from: start, length: fraction)
    }

    /// `Path(roundedRect:)` starts at the top-leading end of the top edge, going clockwise.
    private static func topCenterStartFraction(rect: CGRect, radius: CGFloat) -> CGFloat {
        let straightWidth = max(0, rect.width - 2 * radius)
        let straightHeight = max(0, rect.height - 2 * radius)
        let perimeter = 2 * (straightWidth + straightHeight) + (2 * .pi * radius)
        guard perimeter > 0 else { return 0 }
        return (straightWidth / 2) / perimeter
    }

    private static func trimmedWrapping(_ path: Path, from start: CGFloat, length: CGFloat) -> Path {
        let end = start + length
        if end <= 1 {
            return path.trimmedPath(from: start, to: end)
        }
        var result = path.trimmedPath(from: start, to: 1)
        result.addPath(path.trimmedPath(from: 0, to: end - 1))
        return result
    }
}

File: android/wear/src/main/java/com/wristrally/wear/ScoreScreen.kt
package com.wristrally.wear

import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.wear.compose.material3.MaterialTheme
import androidx.wear.compose.material3.Text
import com.wristrally.domain.MatchService
import com.wristrally.domain.MatchSettings
import com.wristrally.domain.MatchState
import com.wristrally.domain.MatchStatus
import com.wristrally.domain.Side
import com.wristrally.domain.TieBreakNotice
import kotlinx.coroutines.delay

@Composable
fun ScoreScreen(service: MatchService, match: MatchState) {
    val haptic = LocalHapticFeedback.current
    var undoSide by remember { mutableStateOf<Side?>(null) }
    val games = match.scoreScreenSetDisplay
    val game = match.scoreScreenGameDisplay
    val roles = match.servingRoleLabels
    val sides = match.scoreScreenSides

    LaunchedEffect(match.events.size) {
        haptic.performHapticFeedback(HapticFeedbackType.LongPress)
    }
    LaunchedEffect(match.activeTieBreakNotice) {
        if (match.activeTieBreakNotice != null) {
            haptic.performHapticFeedback(HapticFeedbackType.TextHandleMove)
        }
    }
    LaunchedEffect(undoSide) {
        if (undoSide != null) {
            delay((MatchSettings.QUICK_UNDO_TIMEOUT_SECONDS * 1000).toLong())
            undoSide = null
        }
    }

    Column(
        modifier = Modifier.fillMaxSize().padding(horizontal = 6.dp, vertical = 8.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        when {
            match.currentGame.isTieBreak -> TieBreakHeader(match, games)
            match.currentGame.isGoldenPointActive -> DecidingPointHeader(match)
            else -> GamesHeader(match, games)
        }
        Row(
            modifier = Modifier.fillMaxWidth().weight(1f),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            ScoreButton(
                logicalSide = sides.first,
                score = game.first,
                role = roles.first,
                tint = if (sides.first == Side.Left) UsBlue else ThemRed,
                serving = match.currentServer == sides.first,
                undoing = undoSide == sides.first,
                modifier = Modifier.weight(1f).fillMaxHeight(),
                onTap = { handleTap(service, sides.first, undoSide) { undoSide = it } },
            )
            ScoreButton(
                logicalSide = sides.second,
                score = game.second,
                role = roles.second,
                tint = if (sides.second == Side.Left) UsBlue else ThemRed,
                serving = match.currentServer == sides.second,
                undoing = undoSide == sides.second,
                modifier = Modifier.weight(1f).fillMaxHeight(),
                onTap = { handleTap(service, sides.second, undoSide) { undoSide = it } },
            )
        }
    }
}

private fun handleTap(
    service: MatchService,
    logicalSide: Side,
    undoSide: Side?,
    setUndo: (Side?) -> Unit,
) {
    if (undoSide == logicalSide) {
        service.undoLastPoint()
        setUndo(null)
        return
    }
    val previous = service.activeMatch
    service.awardPoint(logicalSide)
    val updated = service.activeMatch
    if (!didPointEndGame(previous, updated)) {
        setUndo(logicalSide)
    } else {
        setUndo(null)
    }
}

private fun didPointEndGame(previous: MatchState?, updated: MatchState?): Boolean {
    if (previous == null || updated == null) return false
    if (previous.status != MatchStatus.InProgress || updated.status != MatchStatus.InProgress) return false
    val previousGames = previous.currentSet.leftGames + previous.currentSet.rightGames
    val updatedGames = updated.currentSet.leftGames + updated.currentSet.rightGames
    if (updatedGames > previousGames) return true
    return updated.completedSets.size > previous.completedSets.size
}

@Composable
private fun TieBreakHeader(match: MatchState, games: Pair<String, String>) {
    val isMatchTB = match.isMatchTieBreak
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Text(
            if (isMatchTB) "Super TB" else "Tie-break",
            color = UndoOrange,
            fontWeight = FontWeight.Bold,
            style = MaterialTheme.typography.labelSmall,
        )
        if (!isMatchTB) {
            Text(
                "${games.first} – ${games.second}",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        val notice = match.activeTieBreakNotice
        Text(
            when (notice) {
                TieBreakNotice.ChangeServe -> "Change serve"
                TieBreakNotice.ChangeSides -> "Change sides"
                null -> if (isMatchTB) "First to 10" else "First to 7"
            },
            style = MaterialTheme.typography.labelSmall,
            color = if (notice != null) UndoOrange else MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

@Composable
private fun GamesHeader(match: MatchState, games: Pair<String, String>) {
    val status = if (match.settings.deuceFormat.numbersDeuceCycles) match.gameStatusLine else null
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Text(
            "${games.first} – ${games.second}",
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.SemiBold,
        )
        if (status != null) {
            Text(
                status,
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

@Composable
private fun DecidingPointHeader(match: MatchState) {
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Text(match.settings.deuceFormat.decidingPointLabel, color = Color(0xFFFFD60A), fontWeight = FontWeight.Bold)
        Text("Next point wins", style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
    }
}

@Composable
private fun ScoreButton(
    logicalSide: Side,
    score: String,
    role: String,
    tint: Color,
    serving: Boolean,
    undoing: Boolean,
    modifier: Modifier,
    onTap: () -> Unit,
) {
    androidx.wear.compose.material3.Button(
        onClick = onTap,
        modifier = modifier.semantics {
            contentDescription = if (serving) "$role $score, serving" else "$role $score"
        },
        colors = androidx.wear.compose.material3.ButtonDefaults.buttonColors(
            containerColor = tint.copy(alpha = if (undoing) 0.45f else 0.28f),
        ),
        shape = WatchButtonShape,
    ) {
        Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Text(score, style = MaterialTheme.typography.displaySmall, fontWeight = FontWeight.Bold, textAlign = TextAlign.Center)
                Text(role, style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
            }
            if (serving) {
                Box(
                    Modifier
                        .align(Alignment.TopCenter)
                        .padding(top = 4.dp)
                        .size(8.dp)
                        .clip(CircleShape)
                        .border(0.dp, Color.Transparent, CircleShape)
                        .then(Modifier),
                ) {
                    Box(Modifier.fillMaxSize().clip(CircleShape).then(Modifier), contentAlignment = Alignment.Center) {
                        androidx.compose.foundation.Canvas(Modifier.size(8.dp)) {
                            drawCircle(Color(0xFFFFD60A))
                        }
                    }
                }
            }
        }
    }
}
